File: go/fixture/go-mini/internal/pkg/deviceid/deviceid.go
// Package deviceid identifies devices.
package deviceid

import (
	"errors"
	"fmt"
	"strings"
)

const maxLen = 64

// DeviceID is the identifier a device reports in its heartbeats.
type DeviceID string

// Parse validates raw and returns it as a DeviceID.
func Parse(raw string) (DeviceID, error) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return "", errors.New("device id: empty")
	}
	if len(raw) > maxLen {
		return "", fmt.Errorf("device id: %d chars, want at most %d", len(raw), maxLen)
	}
	return DeviceID(raw), nil
}

// String returns the id as a string.
func (d DeviceID) String() string {
	return string(d)
}

File: py/fixture/py-mini/internal/pkg/deviceid/deviceid.py
"""The identifier a device reports in its heartbeats."""

from typing import Self

_MAX_LEN = 64


class DeviceID(str):
    """The identifier a device reports in its heartbeats.

    Built only from a non-empty string of at most 64 characters; surrounding
    whitespace is dropped as it is built, so every DeviceID in circulation is
    already trimmed and bounded.
    """

    __slots__ = ()

    def __new__(cls, raw: str) -> Self:
        """Trim raw and reject an empty or over-long identifier."""
        raw = raw.strip()
        if not raw:
            raise ValueError("device id: empty")
        if len(raw) > _MAX_LEN:
            raise ValueError(f"device id: {len(raw)} chars, want at most {_MAX_LEN}")
        return super().__new__(cls, raw)

    @classmethod
    def parse(cls, raw: str) -> Self:
        """Validate raw and return it as a DeviceID."""
        return cls(raw)
